File: TheNexusPlatform/backend-ts/src/claude.ts
import Anthropic from "@anthropic-ai/sdk";
import type { z } from "zod";

import { getSettings } from "./config";

let _client: Anthropic | null = null;

function _getClient(): Anthropic {
  if (_client !== null) return _client;
  const settings = getSettings();
  if (!settings.anthropicApiKey || settings.anthropicApiKey === "your_anthropic_api_key_here") {
    throw new Error("ANTHROPIC_API_KEY is not configured. Set it in backend-ts/.env");
  }
  _client = new Anthropic({ apiKey: settings.anthropicApiKey });
  return _client;
}

/** Single text completion (mirrors the LIAC lib/claude.ts wrapper). */
export async function callClaude(
  systemPrompt: string,
  userMessage: string,
  maxTokens = 3000,
): Promise<string> {
  const client = _getClient();
  const settings = getSettings();
  const message = await client.messages.create({
    model: settings.claudeModel,
    max_tokens: maxTokens,
    system: systemPrompt,
    messages: [{ role: "user", content: userMessage }],
  });
  const block = message.content[0];
  if (!block || block.type !== "text") {
    throw new Error("Unexpected Claude response type");
  }
  return block.text;
}

/** Strip markdown fences / prose and return the JSON object substring. */
export function extractJson(text: string): string {
  let t = text.trim();
  if (t.startsWith("```")) {
    // remove opening fence (```json or ```) and trailing fence
    const parts = t.split("```");
    t = parts.length >= 3 ? parts[1] : t.replace(/`/g, "");
    if (t.trimStart().toLowerCase().startsWith("json")) {
      t = t.trimStart().slice(4);
    }
  }
  const start = t.indexOf("{");
  const end = t.lastIndexOf("}");
  if (start !== -1 && end !== -1 && end > start) {
    return t.slice(start, end + 1);
  }
  return t.trim();
}

/**
 * Call Claude and validate the response against a Zod schema.
 *
 * Retries once (with a corrective nudge) if the first response is not valid
 * JSON matching the schema.
 */
export async function callClaudeJson<S extends z.ZodTypeAny>(
  systemPrompt: string,
  userMessage: string,
  schema: S,
  maxTokens = 3500,
): Promise<z.output<S>> {
  let raw = await callClaude(systemPrompt, userMessage, maxTokens);
  let lastErr: unknown = null;
  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const data = JSON.parse(extractJson(raw));
      const result = schema.safeParse(data);
      if (result.success) return result.data;
      lastErr = result.error;
    } catch (err) {
      lastErr = err;
    }
    if (attempt === 0) {
      raw = await callClaude(
        systemPrompt,
        userMessage +
          "\n\nYour previous response was not valid JSON matching the required shape. " +
          "Return ONLY the raw JSON object, no markdown, no commentary.",
        maxTokens,
      );
    }
  }
  throw new Error(`Claude did not return valid JSON: ${lastErr}`);
}

File: TheNexusPlatform/backend/app/claude.py
import json
from typing import Type, TypeVar

from anthropic import Anthropic
from pydantic import BaseModel, ValidationError

from .config import get_settings

T = TypeVar("T", bound=BaseModel)

_client: Anthropic | None = None


def _get_client() -> Anthropic:
    global _client
    if _client is not None:
        return _client
    settings = get_settings()
    if not settings.anthropic_api_key or settings.anthropic_api_key == "your_anthropic_api_key_here":
        raise RuntimeError("ANTHROPIC_API_KEY is not configured. Set it in backend/.env")
    _client = Anthropic(api_key=settings.anthropic_api_key)
    return _client


def call_claude(system_prompt: str, user_message: str, max_tokens: int = 3000) -> str:
    """Single text completion (mirrors the LIAC lib/claude.ts wrapper)."""
    client = _get_client()
    settings = get_settings()
    message = client.messages.create(
        model=settings.claude_model,
        max_tokens=max_tokens,
        system=system_prompt,
        messages=[{"role": "user", "content": user_message}],
    )
    block = message.content[0]
    if block.type != "text":
        raise RuntimeError("Unexpected Claude response type")
    return block.text


def _extract_json(text: str) -> str:
    """Strip markdown fences / prose and return the JSON object substring."""
    t = text.strip()
    if t.startswith("```"):
        # remove opening fence (```json or ```) and trailing fence
        t = t.split("```", 2)[1] if t.count("```") >= 2 else t.strip("`")
        if t.lstrip().lower().startswith("json"):
            t = t.lstrip()[4:]
    start = t.find("{")
    end = t.rfind("}")
    if start != -1 and end != -1 and end > start:
        return t[start : end + 1]
    return t.strip()


def call_claude_json(system_prompt: str, user_message: str, schema: Type[T], max_tokens: int = 3500) -> T:
    """Call Claude and validate the response against a Pydantic schema.

    Retries once (with a corrective nudge) if the first response is not valid
    JSON matching the schema.
    """
    raw = call_claude(system_prompt, user_message, max_tokens=max_tokens)
    last_err: Exception | None = None
    for attempt in range(2):
        candidate = raw if attempt == 0 else raw
        try:
            data = json.loads(_extract_json(candidate))
            return schema.model_validate(data)
        except (json.JSONDecodeError, ValidationError) as err:
            last_err = err
            if attempt == 0:
                raw = call_claude(
                    system_prompt,
                    user_message
                    + "\n\nYour previous response was not valid JSON matching the required shape. "
                    "Return ONLY the raw JSON object, no markdown, no commentary.",
                    max_tokens=max_tokens,
                )
    raise RuntimeError(f"Claude did not return valid JSON: {last_err}")
